src/index.js

```diff
@@ -1,44 +1,37 @@
-"use strict";
+'use strict';
 
-var isInteger = require("is-integer");
-var matrix = require('ml-matrix');
-
-/**
- * Savitzky-Golay filter
- * @param {Array <number>} data
- * @param {number} h
- * @constructor
- */
-function SavitzkyGolay (data, h) {
-    this.data = data;
-    this.h = h;
-}
+var Matrix = require('ml-matrix');
+var padArray = require('ml-pad-array');
+var extend = require('extend');
 
 var defaultOptions = {
     windowSize: 5,
     derivative: 1,
-    polynomial: 2
+    polynomial: 2,
+    padValue: 0
 };
 
 /**
- * Applies the filter
+ * Savitzky-Golay filter
+ * @param {Array <number>} data
+ * @param {number} h
  * @param {Object} options
  * @returns {Array}
  */
-SavitzkyGolay.prototype.calc = function (options) {
-    options = options || {};
-    for (var o in defaultOptions)
-        if (!(options.hasOwnProperty(o)))
-            options[o] = defaultOptions[o];
-    if ((options.windowSize % 2 === 0) || (options.windowSize < 5) || !(isInteger(options.windowSize)))
+function SavitzkyGolay (data, h, options) {
+    options = extend({}, defaultOptions, options);
+    if ((options.windowSize % 2 === 0) || (options.windowSize < 5) || !(Number.isInteger(options.windowSize)))
         throw new RangeError('Invalid window size (should be odd and at least 5 integer number)');
-    if ((options.derivative < 0) || !(isInteger(options.derivative)))
+    if ((options.derivative < 0) || !(Number.isInteger(options.derivative)))
         throw new RangeError('Derivative should be a positive integer');
-    if ((options.polynomial < 1) || !(isInteger(options.polynomial)))
+    if ((options.polynomial < 1) || !(Number.isInteger(options.polynomial)))
         throw new RangeError('Polynomial should be a positive integer');
 
     var C, norm;
-    var ans =  new Array(this.data.length);
+    var step = Math.floor(options.windowSize / 2);
+    data = padArray(data, {size: step, value: options.padValue});
```

Comment:
Whatever you do to fill up those values, makes a strong assumption on your data, (the other values are 0, or the last known value, or the values that keeps your nth derivative constant on that region), so I think that by default you should lose those points, because you have a different uncertainty on them than in the other true points.